text/0000-alloca.md

```diff
@@ -0,0 +1,147 @@
+- Feature Name: alloca
+- Start Date: 2016-12-01
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a builtin `fn core::mem::reserve<'a, T>(elements: usize) -> StackSlice<'a, T>` that reserves space for the given
+number of elements on the stack and returns a `StackSlice<'a, T>` to it which derefs to `&'a [T]`.
+
+# Motivation
+[motivation]: #motivation
+
+Some algorithms (e.g. sorting, regular expression search) need a one-time backing store for a number of elements only
+known at runtime. Reserving space on the heap always takes a performance hit, and the resulting deallocation can
+increase memory fragmentation, possibly slightly degrading allocation performance further down the road.
+
+If Rust included this zero-cost abstraction, more of these algorithms could run at full speed – and would be available
+on systems without an allocator, e.g. embedded, soft-real-time systems. The option of using a fixed slice up to a
+certain size and using a heap-allocated slice otherwise (as afforded by
+[SmallVec](https://crates.io/crates/smallvec)-like classes) has the drawback of decreasing memory locality if only a
+small part of the fixed-size allocation is used – and even those implementations could potentially benefit from the
+increased memory locality.
+
+As a (flawed) benchmark, consider the following C program:
+
+```C
+#include <stdlib.h>
+
+int main(int argc, char **argv) {
+    int n = argc > 1 ? atoi(argv[0]) : 1;
+    int x = 1;
+    char foo[n];
+    foo[n - 1] = 1;
+}
+```
+
+Running `time nice -n 20 ionice ./dynalloc 1` returns almost instantly (0.0001s), whereas using `time nice -n 20 ionice
+./dynalloc 200000` takes 0.033 seconds. As such, it appears that just by forcing the second write further away from the
+first slows down the program (this benchmark is actually completely unfair, because by reducing the process' priority,
+we invite the kernel to swap in a different process instead, which is very probably the major cause of the slowdown
+here).
+
+Still, even with the flaws in this benchmark,
+[The Myth of RAM](http://www.ilikebigbits.com/blog/2014/4/21/the-myth-of-ram-part-i) argues quite convincingly for the
+benefits of memory frugality.
+
+
+# Detailed design
+[design]: #detailed-design
+
+The standard library function can simply `panic!(..)` within the `reserve(_)` method, as it will be replaced when
+translating to MIR. The `StackSlice` type can be implemented as follows:
+
+```Rust
+/// A slice of data on the stack
+pub struct StackSlice<'a, T: 'a> {
+    slice: &'a [T],
+}
+
+impl<'a, T: 'a> Deref for StackSlice<'a, T> {
+    type Target = [T];
+
+    fn deref(&self) -> &[T] {
+        return self.slice;
+    }
+}
+```
```

Comment:
Is this the full API? Because there seems to be no way to mutate the slice, and no way to initialize it...so as written, this looks like all you can do with it is to read uninitialized memory (!). That memory is from the stack too, which means there could be return addresses...which might help an attacker in some scenarios?

(Well, if `T` has interior mutability then the slice can be mutated, but the uninitialized memory problem still remains.)